File: backend/handlers/todo.go
package handlers

import (
	"net/http"

	"todo-backend/database"
	"todo-backend/models"
	"todo-backend/utils"

	"github.com/gin-gonic/gin"
)


func GetAllTodos(c *gin.Context) {
	query := `SELECT id, title, description, completed, author_id, created_at FROM todos;`
	rows, err := database.DB.Query(c.Request.Context(), query)

	if err != nil {
		utils.SendResponse(
			c,
			http.StatusInternalServerError,
			"Failed to retrieve todos",
			err,
			nil,
		)
		return
	}
	defer rows.Close()

	var todos []models.Todo
	for rows.Next() {
		var todo models.Todo
		if err := rows.Scan(&todo.ID, &todo.Title, &todo.Description, &todo.Completed, &todo.AuthorID, &todo.CreatedAt); err != nil {
			utils.SendResponse(
				c,
				http.StatusInternalServerError,
				"Failed to process todo data",
				err,
				nil,
			)
			return
		}
		todos = append(todos, todo)
	}

	// Ensure that if `todos` is nil, you send an empty array
	if todos == nil {
		todos = []models.Todo{}
	}

	utils.SendResponse(
		c,
		http.StatusOK,
		"Successfully retrieved todo",
		nil,
		todos,
	)
}


func GetTodoByID(c *gin.Context) {
	id := c.Param("id")

	query := `SELECT id, title, description, completed, author_id, created_at FROM todos WHERE id=$1;`
	row := database.DB.QueryRow(c.Request.Context(), query, id)

	var todo models.Todo

	err := row.Scan(&todo.ID, &todo.Title, &todo.Description, &todo.Completed, &todo.AuthorID, &todo.CreatedAt)

	if err != nil {
		utils.SendResponse(
			c,
			http.StatusNotFound,
			"Failed to retrieve todo: todo not found",
			err,
			nil,
		)
		return
	}

	utils.SendResponse(
		c,
		http.StatusOK,
		"Successfully retrieved todo",
		nil,
		todo,
	)
}

func CreateTodo(c *gin.Context) {
	var todo models.Todo

	if err := c.ShouldBindJSON(&todo); err != nil {
		utils.SendResponse(
			c,
			http.StatusBadRequest,
			"Failed to create todo: invalid input",
			err,
			nil,
		)
		return
	}

	query := `INSERT INTO todos (title, description, author_id) VALUES ($1, $2, $3) RETURNING id, completed, created_at;`
	err := database.DB.QueryRow(
		c.Request.Context(),
		query,
		todo.Title,
		todo.Description,
		todo.AuthorID,
	).Scan(&todo.ID, &todo.Completed, &todo.CreatedAt)

	if err != nil {
		utils.SendResponse(
			c,
			http.StatusInternalServerError,
			"Failed to create todo",
			err,
			nil,
		)
		return
	}

	utils.SendResponse(
		c,
		http.StatusCreated,
		"Successfully created todo",
		nil,
		todo,
	)
}


func UpdateTodo(c *gin.Context) {
	id := c.Param("id")
	var todo models.Todo

	if err := c.ShouldBindJSON(&todo); err != nil {
		utils.SendResponse(
			c,
			http.StatusBadRequest,
			"Could not update todo: invalid input",
			err,
			nil,
		)
		return
	}

	query := `UPDATE todos SET title=$1, description=$2, completed=$3, author_id=$4 WHERE id=$5 RETURNING created_at;`
	err := database.DB.QueryRow(
		c.Request.Context(),
		query,
		todo.Title,
		todo.Description,
		todo.Completed,
		todo.AuthorID,
		id,
	).Scan(&todo.ID, &todo.CreatedAt)

	if err != nil {
		utils.SendResponse(
			c,
			http.StatusInternalServerError,
			"Could not update todo",
			err,
			nil,
		)
		return
	}
	
	utils.SendResponse(
		c,
		http.StatusOK,
		"Successfully updated todo",
		nil,
		todo,
	)
}


func DeleteTodo(c *gin.Context) {
	id := c.Param("id")

	query := `DELETE FROM todos WHERE id=$1;`
	_, err := database.DB.Exec(c.Request.Context(), query, id)

	if err != nil {
		utils.SendResponse(c,
			http.StatusInternalServerError,
			"Could not delete todo",
			err,
			false,
		)
		return
	}

	utils.SendResponse(
		c,
		http.StatusOK,
		"Successfully deleted todo",
		nil,
		true,
	)
}

File: backend/src/handlers/todo.rs
use std::{str::FromStr, sync::Arc};

use crate::{
    models::{response::{ApiError, ApiResponse}, todo::{CreateTodo, Todo, UpdateTodo}},
    utils::jwt::{extract_token, validate_jwt},
};
use actix_web::{http::StatusCode, web::{Data, Json, Path}, HttpRequest, HttpResponse};
use deadpool_redis::Pool;
use tokio::sync::Mutex;
use tokio_postgres::Client;
use uuid::Uuid;

static TABLE: &str = "todos";
static FIELDS: &str = "title, description, completed, author_id";
static UPDATE_FIELDS: &str = "title = $1, description = $2, completed = $3, author_id = $4";
static FIELDS_INPUT: &str = "$1, $2, $3, $4";

/// **Create Todo**
pub async fn create_todo(
    client: Data<Arc<Mutex<Client>>>,
    redis_pool: Data<Arc<Pool>>,
    payload: Json<CreateTodo>,
    request: HttpRequest,
) -> Result<HttpResponse, ApiError> {
    let token = extract_token(&request.headers()).map_err(ApiError::from)?;
    let claims = validate_jwt(&token).map_err(ApiError::from)?;
    let user_id = Uuid::from_str(&claims.sub).map_err(ApiError::from)?;

    let client = client.lock().await;
    let stmt = format!(
        "INSERT INTO {} ({}) VALUES ({}) RETURNING *",
        TABLE, FIELDS, FIELDS_INPUT
    );

    let row = client
        .query_one(
            &stmt,
            &[&payload.title, &payload.description, &false, &user_id],
        )
        .await
        .map_err(ApiError::from)?;

    let todo = Todo::from_row(&row);

    // ✅ Store new todo in Redis
    todo.to_redis(&redis_pool, user_id).await?;

    Ok(ApiResponse::success(
        todo,
        "Todo created successfully",
        StatusCode::CREATED,
    ))
}


/// **Read Todo**
pub async fn get_todo(
    client: Data<Arc<Mutex<Client>>>,
    redis_pool: Data<Arc<Pool>>,
    request: HttpRequest,
    id: Path<Uuid>,
) -> Result<HttpResponse, ApiError> {
    let token = extract_token(&request.headers()).map_err(ApiError::from)?;
    let claims = validate_jwt(&token).map_err(ApiError::from)?;
    let id = *id;
    let user_id = Uuid::from_str(&claims.sub).map_err(ApiError::from)?;

    if let Some(todo) = Todo::from_redis(&redis_pool, id, user_id).await? {
        return Ok(ApiResponse::success(todo, "Todo retrieved from cache", StatusCode::OK));
    }

    let client = client.lock().await;
    let stmt = format!("SELECT * FROM {} WHERE id = $1 AND author_id = $2", TABLE);
    let row = client
        .query_one(&stmt, &[&id, &user_id])
        .await
        .map_err(ApiError::from)?;

    let todo = Todo::from_row(&row);

    todo.to_redis(&redis_pool, user_id).await?;

    Ok(ApiResponse::success(
        todo,
        "Todo retrieved successfully",
        StatusCode::OK,
    ))
}

/// **Read Todos**
pub async fn get_todos(
    client: Data<Arc<Mutex<Client>>>,
    redis_pool: Data<Arc<Pool>>,
    request: HttpRequest,
) -> Result<HttpResponse, ApiError> {
    let token = extract_token(&request.headers()).map_err(ApiError::from)?;
    let claims = validate_jwt(&token).map_err(ApiError::from)?;
    let user_id = Uuid::from_str(&claims.sub).map_err(ApiError::from)?;

    let client = client.lock().await;
    let stmt = format!("SELECT * FROM {} WHERE author_id = $1", TABLE);
    let rows = client
        .query(&stmt, &[&user_id])
        .await
        .map_err(ApiError::from)?;

    let mut todos = Vec::new();

    for row in rows {
        let todo = Todo::from_row(&row);
        todo.to_redis(&redis_pool, user_id).await?;
        todos.push(todo);
    }

    Ok(ApiResponse::success(
        todos,
        "Todos retrieved successfully",
        StatusCode::OK,
    ))
}

/// **Update Todo**
pub async fn update_todo(
    client: Data<Arc<Mutex<Client>>>,
    redis_pool: Data<Arc<Pool>>,
    payload: Json<UpdateTodo>,
    request: HttpRequest,
    id: Path<Uuid>,
) -> Result<HttpResponse, ApiError> {
    let token = extract_token(&request.headers()).map_err(ApiError::from)?;
    let claims = validate_jwt(&token).map_err(ApiError::from)?;
    let user_id = Uuid::from_str(&claims.sub).map_err(ApiError::from)?;

    let client = client.lock().await;
    let stmt = format!(
        "UPDATE {} SET {} WHERE id = $5 AND author_id = $4",
        TABLE, UPDATE_FIELDS
    );
    let result = client
        .execute(
            &stmt,
            &[
                &payload.title,
                &payload.description,
                &payload.completed,
                &user_id,
                &*id,
            ],
        )
        .await
        .map_err(ApiError::from)?;

    if result == 1 {
        let updated_todo = Todo {
            id: *id,
            title: payload.title.to_owned(),
            description: payload.description.to_owned(),
            completed: payload.completed,
            author_id: user_id,
        };

        // ✅ Remove old cache
        Todo::delete_from_redis(&redis_pool, *id, user_id).await?;

        // ✅ Store updated todo in cache
        updated_todo.to_redis(&redis_pool, user_id).await?;

        return Ok(ApiResponse::success(
            updated_todo,
            "Todo updated successfully",
            StatusCode::OK,
        ));
    } else if result == 0 {
        return Err(ApiError::Custom("Todo not found on update".to_string()));
    }

    Err(ApiError::Custom("Unexpected update count".to_string()))
}

/// **Delete Todo**
pub async fn delete_todo(
    client: Data<Arc<Mutex<Client>>>,
    redis_pool: Data<Arc<Pool>>,
    request: HttpRequest,
    id: Path<Uuid>,
) -> Result<HttpResponse, ApiError> {
    let token = extract_token(&request.headers()).map_err(ApiError::from)?;
    let claims = validate_jwt(&token).map_err(ApiError::from)?;
    let user_id = Uuid::from_str(&claims.sub).map_err(ApiError::from)?;

    let client = client.lock().await;
    let stmt = format!("DELETE FROM {} WHERE id = $1 AND author_id = $2", TABLE);
    let result = client
        .execute(&stmt, &[&*id, &user_id])
        .await
        .map_err(ApiError::from)?;

    if result == 1 {
        // ✅ Remove from cache
        Todo::delete_from_redis(&redis_pool, *id, user_id).await?;

        return Ok(ApiResponse::success(
            id.to_string(),
            "Todo deleted successfully",
            StatusCode::OK,
        ));
    } else if result == 0 {
        return Err(ApiError::Custom("Todo not found".to_string()));
    }

    Err(ApiError::Custom("Unexpected delete count".to_string()))
}
